src/internal.rs

```diff
@@ -25,15 +44,66 @@ use core::sync::atomic::Ordering;
 use alloc::boxed::Box;
 
 use crossbeam_utils::cache_padded::CachePadded;
+use arrayvec::ArrayVec;
 
 use atomic::Owned;
 use collector::{Handle, Collector};
 use epoch::{AtomicEpoch, Epoch};
 use guard::{unprotected, Guard};
-use garbage::{Bag, Garbage};
+use deferred::Deferred;
 use sync::list::{List, Entry, IterError, IsElement};
 use sync::queue::Queue;
 
+/// Maximum number of objects a bag can contain.
+#[cfg(not(feature = "strict_gc"))]
+const MAX_OBJECTS: usize = 64;
+#[cfg(feature = "strict_gc")]
+const MAX_OBJECTS: usize = 4;
+
+/// Bag of garbages.
+#[derive(Default, Debug)]
+pub struct Bag {
+    /// Stashed objects.
+    deferreds: ArrayVec<[Deferred; MAX_OBJECTS]>,
+}
+
+// It is safe to implement `Send` for `Bag`, since the user of `Bag::try_push()` guarantees that
+// another thread may execute any `Deferred` functions inside a `Bag`.
+unsafe impl Send for Bag {}
+
+impl Bag {
+    /// Returns a new, empty bag.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Returns `true` if the bag is empty.
+    pub fn is_empty(&self) -> bool {
+        self.deferreds.is_empty()
+    }
+
+    /// Attempts to insert a garbage object into the bag.
+    ///
+    /// Returns `Ok(())` if successful, and `Err(deferred)` for the given `deferred` if
+    /// unsuccessful.
+    ///
+    /// # Safety
+    ///
+    /// Another thread may execute the `deferred` function.
+    pub unsafe fn try_push(&mut self, deferred: Deferred) -> Result<(), Deferred> {
+        self.deferreds.try_push(deferred).map_err(|e| e.element())
+    }
+}
+
+impl Drop for Bag {
+    fn drop(&mut self) {
+        // Call all deferred functions.
+        for deferred in self.deferreds.iter_mut() {
```

Comment:
I can't see anywhere where that could happen, but are we absolutely sure a `Deferred` is never dropped as a standalone object rather than when it's inside a `Bag`? This change would then make it never run, as opposed to having the individual `Garbage` destructor take care of it before the change.

Reply:
I revised the doc.  Would you please review again to check if it's clearer?

Reply:
Thank you for revising, but I think that I had something slightly different in mind. Basically, after this change, `Deferred` relies on being inside a `Bag` to be destroyed and have its stored function run. On its own, a `Deferred` won't run when dropped. I think that's fine, but just wanted to make sure.

Reply:
I agree with @Vtec234, so I just changed `Deferred::call()` so that it consumes `self`.  Basically `Deferred`'s API is almost the same with `FnOnce<()>`, and so we don't need to worry about double calling.

cc @stjepang, as he is the original author of `Deferred`.  May I ask why did you allow to call `Deferred` twice (and panic)?

Reply:
Maybe at the time we had to be able to do `.call()` without consuming the `Deferred`? I don't remember the details.

If you can change `call` to consume `self`, go ahead. That'd be nicer than what we currently have. 👍